src/Components/Server/src/Circuits/CircuitRegistry.cs

```diff
@@ -0,0 +1,262 @@
+                }
+                else
+                {
+                    // DisconnectCore may fail to disconnect the circuit if it was previously marked inactive or
+                    // has been transfered to a new connection. Do not invoke the circuit handlers in this instance.
+
+                    // We have to do in this instance.
+                    return Task.CompletedTask;
+                }
+            }
+
+            return circuitHandlerTask;
+        }
+
+        protected virtual bool DisconnectCore(CircuitHost circuitHost, string connectionId)
+        {
+            if (!ConnectedCircuits.TryGetValue(circuitHost.CircuitId, out circuitHost))
+            {
+                // Guard: The circuit might already have been marked as inactive.
+                return false;
+            }
+
+            if (!string.Equals(circuitHost.Client.ConnectionId, connectionId, StringComparison.Ordinal))
+            {
+                // The circuit is associated with a different connection. One way this could happen is when
+                // the client reconnects with a new connection before the OnDisconnect for the older
+                // connection is executed. Do nothing
+                return false;
+            }
+
+            var result = ConnectedCircuits.TryRemove(circuitHost.CircuitId, out circuitHost);
+            Debug.Assert(result, "This operation operates inside of a lock. We expect the previously inspected value to be still here.");
+
+            circuitHost.Client.SetDisconnected();
+            var entryOptions = new MemoryCacheEntryOptions
+            {
+                AbsoluteExpiration = DateTimeOffset.UtcNow.Add(_options.DisconnectedCircuitRetentionPeriod),
+                Size = 1,
+                PostEvictionCallbacks = { _postEvictionCallback },
+            };
+
+            DisconnectedCircuits.Set(circuitHost.CircuitId, circuitHost, entryOptions);
+            return true;
+        }
+
+        public virtual async Task<CircuitHost> ConnectAsync(string circuitId, IClientProxy clientProxy, string connectionId, CancellationToken cancellationToken)
+        {
+            CircuitHost circuitHost;
+            bool previouslyConnected;
+
+            Task circuitHandlerTask;
+
+            lock (CircuitRegistryLock)
+            {
+                // Transition the host from disconnected to connected if it's available. In this critical section, we return
+                // an existing host if it's currently considered connected or transition a disconnected host to connected.
+                // Transfering also wires up the client to the new set.
+                (circuitHost, previouslyConnected) = ConnectCore(circuitId, clientProxy, connectionId);
+
+                if (circuitHost == null)
+                {
+                    // Failed to connect. Nothing to do here.
+                    return null;
```

Comment:
null task :( that's pretty unconventional.

Reply:
It's async method, so it's a null value (not `null` Task).